doc-content/drools-docs/src/main/asciidoc/DecisionEngine/phreak-rule-evaluation-con.adoc

```diff
@@ -0,0 +1,68 @@
+[id='phreak-rule-evaluation-con_{context}']
+
+= Rule evaluation in Phreak
+
+When the {DECISION_ENGINE} starts, all rules are considered to be _unlinked_ from pattern-matching data that can trigger the rules. At this stage, the Phreak algorithm in the {DECISION_ENGINE} does not evaluate the rules. The insert, update, and delete actions are queued before entering the beta network. Phreak uses a heuristic, based on the rule most likely to result in executions, to calculate and select the next rule for evaluation. This delays the evaluation and execution of the other rules. When all the required input values are populated for a rule, the rule is considered to be _linked_ to the relevant pattern-matching data. Phreak then creates a goal that represents this rule and places the goal into a priority queue, which is ordered by salience. Each queue is associated with an `AgendaGroup`. Phreak only evaluates rules for the active `AgendaGroup` by inspecting the queue and fulfilling the goal for the rule with the highest salience. This means the work shifts from the insert, update, and delete phase to the `fireAllRules` phase. Only the rule for which the goal was created is evaluated, and other potential rule evaluations are delayed. While individual rules are evaluated, node sharing is still achieved through the process of segmentation.
+
+Unlike the tuple-oriented Rete, the Phreak propagation is collection oriented. For the rule that is being evaluated, the {DECISION_ENGINE} accesses the first node and processes all queued insert, update, and delete actions. The results are added to a set, and the set is propagated to the child node. In the child node, all queued insert, update, and delete actions are processed, adding the results to the same set. The set is then propagated to the next child node and the same process repeats until it reaches the terminal node. This cycle creates a batch process effect that can provide performance advantages for certain rule constructs.
+
+This linking and unlinking of rules happens through a layered bit-mask system, based on network segmentation. When the rule network is built, segments are created for nodes that are shared by the same set of rules. A rule is composed of a path of segments. In case a rule does not share any node with any other rule, it becomes a single segment.
+
+A bit-mask offset is assigned to each node in the segment. Another bit mask is assigned to each segment in the path of the rule according to these requirements:
```

Comment:
Right, but it's best to precede a bulleted list with text more specific to it, rather than a whole paragraph followed by bullets that apply to only the last couple of sentences.